File: android/app/src/main/java/ch/vbckantibaden/android/core/NotificationCenterImpl.java
package ch.vbckantibaden.android.core;

import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

import ch.vbckantibaden.android.listener.ModelListener;
import ch.vbckantibaden.android.listener.PlayerListener;

public class NotificationCenterImpl implements NotificationCenter {

    private static final class InstanceHolder {
        static final NotificationCenterImpl INSTANCE = new NotificationCenterImpl();
    }

    public static NotificationCenterImpl getInstance () {
        return InstanceHolder.INSTANCE;
    }

    private Map<ModelObserverKey, ModelObserver> observers = new ConcurrentHashMap<>();

    private NotificationCenterImpl () {}

    @Override
    public void notify(int resourceId, int key) {
        for (ModelObserver observer : observers.values()) {
            if (observer.isInterested(resourceId, key)) {
                observer.notifyListener();
            }
        }
    }

    @Override
    public void notifyList(int resourceId) {

    }

    public void addPlayerListener(final PlayerListener listener, final Player player) {
        ModelObserverKey observerKey = new ModelObserverKey(listener, player.resourceId(), player.key());
        observers.put(observerKey, new ModelObserver(observerKey) {
            @Override
            protected void notifyListener() {
                listener.playerChanged(player);
            }
        });
    }

    public void removePlayerListener(final PlayerListener listener, final Player player) {
        removeModelListener(listener, player.resourceId(), player.key());
    }

    private void removeModelListener(ModelListener listener, int resourceId, int key) {
        ModelObserverKey observerKey = new ModelObserverKey(listener, resourceId, key);
        observers.remove(observerKey);
    }

    private class ModelObserverKey {
        private ModelListener listener;
        private int resourceId;
        private int key;

        public ModelObserverKey(ModelListener listener, int resourceId, int key) {
            this.listener = listener;
            this.resourceId = resourceId;
            this.key = key;
        }

        @Override
        public boolean equals(Object o) {
            if (o == null || !(o instanceof ModelObserverKey)) {
                return false;
            }
            ModelObserverKey other = (ModelObserverKey)o;
            return listener.equals(other.listener) && resourceId == other.resourceId && key == other.key;
        }

        @Override
        public int hashCode() {
            int hash = listener.hashCode();
            hash += 31 * hash + resourceId;
            return 31 * hash + key;
        }
    }

    private abstract class ModelObserver {
        protected ModelObserverKey key;

        protected ModelObserver(ModelObserverKey key) {
            this.key = key;
        }

        protected boolean isInterested(int resourceId, int key) {
            return resourceId == this.key.resourceId && key == this.key.key;
        }

        protected abstract void notifyListener();
    }
}

File: ios/app/Classes/CoreWrappers/NotificationCenterImpl.m
//
//  NotificationCenterImpl.m
//  app
//
//  Created by Adrian Friedli on 03.01.15.
//  Copyright (c) 2015 Adrian Friedli. All rights reserved.
//

#import "NotificationCenterImpl.h"
#import "ModelObserver.h"
#import "ModelObserverKey.h"
#import "Player.h"
#import "PlayerListener.h"

@implementation NotificationCenterImpl {
	NSMutableDictionary* _observers;
}

#pragma mark - Singleton
+ (instancetype)getInstance {
	static NotificationCenterImpl* instance = nil;
	static dispatch_once_t onceToken;
	dispatch_once(&onceToken, ^{
		instance = [[self alloc] init];
	});
	return instance;
}

#pragma mark - Object Lifecycle
- (instancetype)init {
	self = [super init];
	if (self) {
		_observers = [[NSMutableDictionary alloc] init];
	}
	return self;
}

#pragma mark - Listeners
- (void)addPlayerListener:(id<PlayerListener>)listener player:(Player *)player {
	ModelObserverKey* key = [[ModelObserverKey alloc] initWithListener:listener resourceId:player.resourceId key:player.key];
	[_observers setObject:[[ModelObserver alloc] initWithKey:key notification:^{
		[listener playerChanged:player];
	}] forKey:key];
}

- (void)removePlayerListener:(id<PlayerListener>)listener player:(Player *)player {
	ModelObserverKey* key = [[ModelObserverKey alloc] initWithListener:listener resourceId:player.resourceId key:player.key];
	[_observers removeObjectForKey:key];
}

#pragma mark - NotificationCenter
- (void)notifyListWithResourceId:(int32_t)resourceId {
	
}

- (void)notifyWithResourceId:(int32_t)resourceId key:(int32_t)key {
	for (ModelObserver* observer in [_observers allValues]) {
		if ([observer isInterestedForResourceId:resourceId key:key]) {
			[observer notifyListener];
		}
	}
}



@end
